File: sample-code/examples/python/android_complex.py
import os
import unittest

from appium import webdriver
from appium.webdriver.common.touch_action import TouchAction
from appium.webdriver.common.multi_action import MultiAction

from time import sleep

# Returns abs path relative to this file and not cwd
PATH = lambda p: os.path.abspath(
    os.path.join(os.path.dirname(__file__), p)
)

class ComplexAndroidTests(unittest.TestCase):
    def setUp(self):
        desired_caps = {}
        desired_caps['platformName'] = 'Android'
        desired_caps['platformVersion'] = '4.2'
        desired_caps['deviceName'] = 'Android Emulator'
        desired_caps['app'] = PATH(
            '../../../sample-code/apps/ApiDemos/bin/ApiDemos-debug.apk'
        )

        self.driver = webdriver.Remote('http://localhost:4723/wd/hub', desired_caps)

    def tearDown(self):
        self.driver.quit()

    def test_find_elements(self):
        # pause a moment, so xml generation can occur
        sleep(2)

        els = self.driver.find_elements_by_xpath('//android.widget.TextView')
        self.assertEqual('API Demos', els[0].text)

        el = self.driver.find_element_by_xpath('//android.widget.TextView[contains(@text, "Animat")]')
        self.assertEqual('Animation', el.text)

        el = self.driver.find_element_by_accessibility_id("App")
        el.click()

        els = self.driver.find_elements_by_android_uiautomator('new UiSelector().clickable(true)')
        # there are more, but at least 10 visible
        self.assertLess(10, len(els))
        # the list includes 2 before the main visible elements
        self.assertEqual('Action Bar', els[2].text)

        els = self.driver.find_elements_by_xpath('//android.widget.TextView')
        self.assertLess(10, len(els))
        self.assertEqual('Action Bar', els[1].text)

    def test_scroll(self):
        sleep(2)
        els = self.driver.find_elements_by_xpath('//android.widget.TextView')
        self.driver.scroll(els[7], els[3])

        el = self.driver.find_element_by_accessibility_id('Views')

    def test_smiley_face(self):
        # just for the fun of it.
        # this doesn't really assert anything.
        self.driver.find_element_by_accessibility_id('Graphics').click()

        els = self.driver.find_elements_by_class_name('android.widget.TextView')
        self.driver.scroll(els[len(els)-1], els[0])

        el = None
        try:
            el = self.driver.find_element_by_accessibility_id('Touch Paint')
        except Exception as e:
            els = self.driver.find_elements_by_class_name('android.widget.TextView')
            self.driver.scroll(els[len(els)-1], els[0])

        if el is None:
            el = self.driver.find_element_by_accessibility_id('Touch Paint')

        el.click()

        # paint
        e1 = TouchAction()
        e1.press(x=150, y=100).release()

        e2 = TouchAction()
        e2.press(x=250, y=100).release()

        smile = TouchAction()
        smile.press(x=110, y=200) \
            .move_to(x=1, y=1) \
            .move_to(x=1, y=1) \
            .move_to(x=1, y=1) \
            .move_to(x=1, y=1) \
            .move_to(x=1, y=1) \
            .move_to(x=2, y=1) \
            .move_to(x=2, y=1) \
            .move_to(x=2, y=1) \
            .move_to(x=2, y=1) \
            .move_to(x=2, y=1) \
            .move_to(x=3, y=1) \
            .move_to(x=3, y=1) \
            .move_to(x=3, y=1) \
            .move_to(x=3, y=1) \
            .move_to(x=3, y=1) \
            .move_to(x=4, y=1) \
            .move_to(x=4, y=1) \
            .move_to(x=4, y=1) \
            .move_to(x=4, y=1) \
            .move_to(x=4, y=1) \
            .move_to(x=5, y=1) \
            .move_to(x=5, y=1) \
            .move_to(x=5, y=1) \
            .move_to(x=5, y=1) \
            .move_to(x=5, y=1) \
            .move_to(x=5, y=0) \
            .move_to(x=5, y=0) \
            .move_to(x=5, y=0) \
            .move_to(x=5, y=0) \
            .move_to(x=5, y=0) \
            .move_to(x=5, y=0) \
            .move_to(x=5, y=0) \
            .move_to(x=5, y=0) \
            .move_to(x=5, y=-1) \
            .move_to(x=5, y=-1) \
            .move_to(x=5, y=-1) \
            .move_to(x=5, y=-1) \
            .move_to(x=5, y=-1) \
            .move_to(x=4, y=-1) \
            .move_to(x=4, y=-1) \
            .move_to(x=4, y=-1) \
            .move_to(x=4, y=-1) \
            .move_to(x=4, y=-1) \
            .move_to(x=3, y=-1) \
            .move_to(x=3, y=-1) \
            .move_to(x=3, y=-1) \
            .move_to(x=3, y=-1) \
            .move_to(x=3, y=-1) \
            .move_to(x=2, y=-1) \
            .move_to(x=2, y=-1) \
            .move_to(x=2, y=-1) \
            .move_to(x=2, y=-1) \
            .move_to(x=2, y=-1) \
            .move_to(x=1, y=-1) \
            .move_to(x=1, y=-1) \
            .move_to(x=1, y=-1) \
            .move_to(x=1, y=-1) \
            .move_to(x=1, y=-1)
        smile.release()

        ma = MultiAction(self.driver)
        ma.add(e1, e2, smile)
        ma.perform()

        # so you can see it
        sleep(10)


if __name__ == '__main__':
    suite = unittest.TestLoader().loadTestsFromTestCase(ComplexAndroidTests)
    unittest.TextTestRunner(verbosity=2).run(suite)

File: sample-code/examples/node/android-complex.js
"use strict";

require("./helpers/setup");

var wd = require("wd"),
    _ = require('underscore'),
    actions = require("./helpers/actions"),
    serverConfigs = require('./helpers/appium-servers'),
    _p = require('./helpers/promise-utils'),
    Q = require('q');

wd.addPromiseChainMethod('swipe', actions.swipe);

describe("android complex", function () {
  this.timeout(300000);
  var driver;
  var allPassed = true;

  before(function () {
    var serverConfig = process.env.SAUCE ?
      serverConfigs.sauce : serverConfigs.local;
    driver = wd.promiseChainRemote(serverConfig);
    require("./helpers/logging").configure(driver);

    var desired = process.env.SAUCE ?
      _.clone(require("./helpers/caps").android18) :
      _.clone(require("./helpers/caps").android19);
    desired.app = require("./helpers/apps").androidApiDemos;
    if (process.env.SAUCE) {
      desired.name = 'android - complex';
      desired.tags = ['sample'];
    }
    return driver
      .init(desired)
      .setImplicitWaitTimeout(5000);
  });

  after(function () {
    return driver
      .quit()
      .finally(function () {
        if (process.env.SAUCE) {
          return driver.sauceJobStatus(allPassed);
        }
      });
  });

  afterEach(function () {
    allPassed = allPassed && this.currentTest.state === 'passed';
  });

  it("should find an element", function () {
    return driver
      .elementByXPath('//android.widget.TextView[@text=\'Animation\']')
      .elementByXPath('//android.widget.TextView')
        .text().should.become('API Demos')
      .elementsByXPath('//android.widget.TextView[contains(@text, "Animat")]')
        .then(_p.filterDisplayed).first()
      .then(function (el) {
        if (!process.env.SAUCE) {
          return el.text().should.become('Animation');
        }
      }).elementByName('App').click()
        .sleep(3000)
      .elementsByAndroidUIAutomator('new UiSelector().clickable(true)')
        .should.eventually.have.length.above(10)
      .elementByXPath('//android.widget.TextView[@text=\'Action Bar\']')
        .should.eventually.exist
      .elementsByXPath('//android.widget.TextView')
        .then(_p.filterDisplayed).first()
        .text().should.become('API Demos')
      .back().sleep(1000);
  });

  it("should scroll", function () {
    return driver
      .elementByXPath('//android.widget.TextView[@text=\'Animation\']')
      .elementsByXPath('//android.widget.TextView')
      .then(function (els) {
        return Q.all([
          els[7].getLocation(),
          els[3].getLocation()
        ]).then(function (locs) {
          console.log('locs -->', locs);
          return driver.swipe({
            startX: locs[0].x, startY: locs[0].y,
            endX: locs[1].x, endY: locs[1].y,
            duration: 800
          });
        });
      });
  });

  it("should draw a smiley", function () {
    function findTouchPaint() {
      return driver
        .elementsByClassName('android.widget.TextView')
        .then(function (els) {
          return Q.all([
            els[els.length - 1].getLocation(),
            els[0].getLocation()
          ]).then(function (locs) {
            return driver.swipe({
              startX: locs[0].x, startY: locs[0].y,
              endX: locs[1].x, endY: locs[1].y,
              duration: 800
            });
          });
        }).elementByName('Touch Paint')
        .catch(function () {
          return findTouchPaint();
        });
    }

    return driver
      .elementByName('Graphics').click()
      .then(findTouchPaint)
      .click()
      .sleep(5000)
      .then(function () {
        var a1 = new wd.TouchAction();
        a1.press({x: 150, y: 100}).release();
        var a2 = new wd.TouchAction();
        a2.press({x: 250, y: 100}).release();
        var smile = new wd.TouchAction();
        smile
          .press({x:110, y:200})
          .moveTo({x:1, y:1})
          .moveTo({x:1, y:1})
          .moveTo({x:1, y:1})
          .moveTo({x:1, y:1})
          .moveTo({x:1, y:1})
          .moveTo({x:2, y:1})
          .moveTo({x:2, y:1})
          .moveTo({x:2, y:1})
          .moveTo({x:2, y:1})
          .moveTo({x:2, y:1})
          .moveTo({x:3, y:1})
          .moveTo({x:3, y:1})
          .moveTo({x:3, y:1})
          .moveTo({x:3, y:1})
          .moveTo({x:3, y:1})
          .moveTo({x:4, y:1})
          .moveTo({x:4, y:1})
          .moveTo({x:4, y:1})
          .moveTo({x:4, y:1})
          .moveTo({x:4, y:1})
          .moveTo({x:5, y:1})
          .moveTo({x:5, y:1})
          .moveTo({x:5, y:1})
          .moveTo({x:5, y:1})
          .moveTo({x:5, y:1})
          .moveTo({x:5, y:0})
          .moveTo({x:5, y:0})
          .moveTo({x:5, y:0})
          .moveTo({x:5, y:0})
          .moveTo({x:5, y:0})
          .moveTo({x:5, y:0})
          .moveTo({x:5, y:0})
          .moveTo({x:5, y:0})
          .moveTo({x:5, y:-1})
          .moveTo({x:5, y:-1})
          .moveTo({x:5, y:-1})
          .moveTo({x:5, y:-1})
          .moveTo({x:5, y:-1})
          .moveTo({x:4, y:-1})
          .moveTo({x:4, y:-1})
          .moveTo({x:4, y:-1})
          .moveTo({x:4, y:-1})
          .moveTo({x:4, y:-1})
          .moveTo({x:3, y:-1})
          .moveTo({x:3, y:-1})
          .moveTo({x:3, y:-1})
          .moveTo({x:3, y:-1})
          .moveTo({x:3, y:-1})
          .moveTo({x:2, y:-1})
          .moveTo({x:2, y:-1})
          .moveTo({x:2, y:-1})
          .moveTo({x:2, y:-1})
          .moveTo({x:2, y:-1})
          .moveTo({x:1, y:-1})
          .moveTo({x:1, y:-1})
          .moveTo({x:1, y:-1})
          .moveTo({x:1, y:-1})
          .moveTo({x:1, y:-1})
          .release();

        var ma = new wd.MultiAction().add(a1, a2, smile);
        return driver.performMultiAction(ma)
          // so you can see it
          .sleep(10000)
          .back().sleep(1000)
          .back().sleep(1000);
      });
  });

});
